### PR TITLE
JBR-7572 Vulkan: Implement vertex buffer pool.

### PR DESCRIPTION
**Pull request by YaaZ**
Track and reuse vertex buffers, no need to allocate and bind a new buffer on each draw.

### CODE REVIEW

**Comment by YaaZ** [comment]
High-level overview:
1. We allocate fixed-size (currently 128KiB, in 1MiB pages) vertex buffers on demand.
2. Vertex buffer is taken from the pool or allocated, and then bound to the command buffer when we start drawing.
3. Adding vertices just increments vertex counter, accumulated vertices are flushed with `vkCmdDraw` only when needed (pipeline switch, render pass end, buffer overflow).
4. When vertex format changes, the same buffer is re-bound to the command buffer at a new offset.
5. When (if) current vertex buffer runs out of space, new one is taken.
6. When drawing commands are submitted for execution, filled vertex buffers are returned to the pool and become available for reuse only after corresponding drawing commands complete execution.
7. Same logic can (and will) be reused with other buffer types.

**Comment by bourgesl** [comment]
Sorry no time yet...

**Review by avu — APPROVED** [review]
Overall LGTM, let's see it in action

**Comment by bourgesl** [comment]
I will look at this patch tomorrow morning.

File: src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c [hunk 1]
```diff
@@ -34,14 +34,51 @@
...
+// In debug mode resource reuse will be randomly delayed by 3 timestamps in ~20% cases.
+#define POOL_RETURN(RENDERER, NAME, VAR) RING_BUFFER_PUSH_BACK((RENDERER)->NAME, \
+    (struct PoolEntry_ ## NAME) { .timestamp = (RENDERER)->writeTimestamp + (VK_DEBUG_RANDOM(20)*3), .value = (VAR) })
+
+/**
+ * Insert an item into the pool. It is available for POOL_TAKE immediately.
+ * This is usually used for bulk insertion of newly-created resources.
+ */
+#define POOL_INSERT(RENDERER, NAME, VAR) RING_BUFFER_PUSH_FRONT((RENDERER)->NAME, \
+    (struct PoolEntry_ ## NAME) { .timestamp = 0ULL, .value = (VAR) })
```
Comment (bourgesl) [inline src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c:68]: 0ULL !!
0 , not NULL ?

Comment (YaaZ) [reply src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c:68]: It's intended, 0 as unsigned long long

File: src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c [hunk 2]
```diff
@@ -198,24 +274,26 @@ VKRenderer* VKRenderer_Create(VKDevice* device) {
...
-        renderer->device->vkDestroySemaphore(renderer->device->handle, renderer->timelineSemaphore, NULL);
-    }
-    if (renderer->commandPool != VK_NULL_HANDLE) {
-        renderer->device->vkDestroyCommandPool(renderer->device->handle, renderer->commandPool, NULL);
+    POOL_FREE(renderer, commandBufferPool);
+    POOL_FREE(renderer, secondaryCommandBufferPool);
+    POOL_DRAIN(renderer, semaphorePool, entry) device->vkDestroySemaphore(device->handle, entry->value, NULL);
+
+    // Destroy vertex buffer pool.
+    POOL_DRAIN(renderer, vertexBufferPool, entry) device->vkDestroyBuffer(device->handle, entry->value.handle, NULL);
```
Comment (bourgesl) [inline src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c:289]: please use braces here to understand POOL_DRAIN creates a FOR loop => rename to POOL_DRAIN_FOR() { ... }

File: src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c [hunk 3]
```diff
@@ -198,24 +274,26 @@ VKRenderer* VKRenderer_Create(VKDevice* device) {
...
-    }
-    RING_BUFFER_FREE(renderer->pendingSemaphores);
-    if (renderer->timelineSemaphore != VK_NULL_HANDLE) {
-        renderer->device->vkDestroySemaphore(renderer->device->handle, renderer->timelineSemaphore, NULL);
-    }
-    if (renderer->commandPool != VK_NULL_HANDLE) {
-        renderer->device->vkDestroyCommandPool(renderer->device->handle, renderer->commandPool, NULL);
+    POOL_FREE(renderer, commandBufferPool);
+    POOL_FREE(renderer, secondaryCommandBufferPool);
+    POOL_DRAIN(renderer, semaphorePool, entry) device->vkDestroySemaphore(device->handle, entry->value, NULL);
```
Comment (bourgesl) [inline src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c:286]: same

File: src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c (hunk 1 above)
Comment (bourgesl) [reply src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c:68]: Or 0ull to understand it is 0

Comment (tsarn) [reply src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c:68]: Or if we're being really pedantic, `UINT64_C(0)`, for those hypothetical platforms where `sizeof(unsigned long long) != 8`. But I think it really doesn't matter.

Comment (YaaZ) [reply src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c:68]: Such a hot topic, maybe just 0 would be good enough? 😀

File: src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c [hunk 4]
```diff
@@ -34,14 +34,51 @@
...
+        (VAR) = RING_BUFFER_FRONT((RENDERER)->NAME)->value; RING_BUFFER_POP_FRONT((RENDERER)->NAME); \
+    }} while(0)
+
+/**
+ * Return an item to the pool. It will only become available again
+ * after the next submitted batch of work completes execution on GPU.
+ */
+// In debug mode resource reuse will be randomly delayed by 3 timestamps in ~20% cases.
+#define POOL_RETURN(RENDERER, NAME, VAR) RING_BUFFER_PUSH_BACK((RENDERER)->NAME, \
+    (struct PoolEntry_ ## NAME) { .timestamp = (RENDERER)->writeTimestamp + (VK_DEBUG_RANDOM(20)*3), .value = (VAR) })
```
Comment (bourgesl) [inline src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c:61]: are you sure that compiler can remove (0 * 3) if not DEBUG set and VK_DEBUG_RANDOM = 0 ?

Comment (mkartashev) [reply src/java.desktop/share/native/common/java2d/vulkan/VKRenderer.c:61]: I can vouch for this.

**Review by bourgesl — APPROVED** [review]
LGTM, only minor syntax change (braces arround for) but not mandatory

**Comment by YaaZ** [comment]
Fixed remaining remarks in `main`

**Pull request closed without merging**